langchain/src/callbacks/tracers.ts

```diff
@@ -98,43 +97,45 @@ export abstract class BaseTracer extends BaseCallbackHandler {
     }
   }
 
-  protected _startTrace(runId: string, run: LLMRun | ChainRun | ToolRun) {
+  protected _startTrace(run: LLMRun | ChainRun | ToolRun) {
     this.executionOrder += 1;
 
-    if (this.stack.length > 0) {
-      if (
-        !(
-          this.stack.at(-1)?.type === "tool" ||
-          this.stack.at(-1)?.type === "chain"
-        )
-      ) {
-        throw new Error("Nested run can only be logged for tool or chain");
+    if (run.parent_uuid) {
+      const parentRun = this.runMap.get(run.parent_uuid);
+      if (parentRun) {
+        if (!(parentRun.type === "tool" || parentRun.type === "chain")) {
+          throw new Error("Caller run can only be a tool or chain");
+        } else {
+          this._addChildRun(parentRun as ChainRun | ToolRun, run);
+        }
+      } else {
+        throw new Error(`Caller run ${run.parent_uuid} not found`);
       }
-      const parentRun = this.stack.at(-1) as ChainRun | ToolRun;
-      this._addChildRun(parentRun, run);
     }
-    this.runMap.set(runId, run);
-    this.stack.push(run);
+    this.runMap.set(run.uuid, run);
   }
 
-  protected async _endTrace(runId: string) {
-    const run = this.runMap.get(runId);
-    this.stack.pop();
-    if (this.stack.length === 0 && run) {
-      this.executionOrder = 1;
+  protected async _endTrace(run: LLMRun | ChainRun | ToolRun): Promise<void> {
+    if (!run.parent_uuid) {
       await this.persistRun(run);
+      this.executionOrder = 1;
```

Comment:
Do we still need this? Sounds like it can instead be a calculated field?

Reply:
Potentially not, but would rather not change this for now as it's something our backend uses atm